# **Ejercicios de pair programming Módulo 3 Sprint 1**
## **Regreción Lineal: Lección 7 - ANOVA**

En el pair programming de hoy usaremos el set de datos que guardastéis en el pair programming de normalización y estandarización.


Hasta ahora habéis estado evaluando las características de vuestro set de datos y habéis hecho una gran exploración, es el momento de hacer vuestro primer ANOVA! En el ejercicio de hoy tendréis que hacer un ANOVA con vuestro datos y hacer una interpretación de los resultados.


📌 NOTA Puede que vuestros datos no se ajusten o no cumplan todas las asunciones, no pasa nada, haced el ANOVA e interpretad los resultados. En próximas lecciones aprenderemos que podemos hacer cuando nos encontramos en esta situación.

In [1]:
# Tratamiento de los datos
# ========================
import numpy as np
import pandas as pd

# Librerías para la visualización de los datos
# ============================================ 
import matplotlib.pyplot as plt 
import seaborn as sns

# Estadísticos
# ============
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.multivariate.manova import MANOVA
from sklearn.preprocessing import StandardScaler

# Establecer preferencias de visualización
# ========================================
plt.rcParams["figure.figsize"] = (10,8) 

In [2]:
df = pd.read_csv("datos/soci_econ_country_profiles_estandarizado.csv", index_col = 0)
df.sample(5)

region  restaurant_price_index  population_2017  \
country                                                                    
Vietnam       South-easternAsia                   20.46         1.305612   
Argentina          SouthAmerica                   50.37         0.432582   
Saudi Arabia        WesternAsia                   30.22         0.239602   
Hungary           EasternEurope                   32.72        -0.155722   
Japan               EasternAsia                   47.83         1.849540   

              employment_agriculture  employment_industry  \
country                                                     
Vietnam                     2.964803            -0.131944   
Argentina                  -0.331263             0.131944   
Saudi Arabia               -0.008282            -0.159722   
Hungary                    -0.140787             0.826389   
Japan                      -0.190476             0.368056   

              employment_services  education  quality_of_life_index  \
country                                                               
Vietnam                 -1.859397   0.571429              -1.628793   
Argentina                0.315638   0.285714              -0.022740   
Saudi Arabia             0.218077   0.144640               0.083120   
Hungary                 -0.097561  -0.142857              -0.034692   
Japan                    0.126255  -0.928571               0.099884   

              purchasing_power_index  safety_index  cost_of_living  \
country                                                              
Vietnam                    -0.941793     -0.723245       -0.280420   
Argentina                  -0.243315     -1.258757        0.080347   
Saudi Arabia                1.410977      0.899903       -0.041260   
Hungary                    -0.469287      0.083622       -0.203402   
Japan                       0.673036      1.052748        1.027434   

              consumer_price_index  human_capital_index  inflation  \
country                                                              
Vietnam                   1.982659            -0.246291   0.689243   
Argentina                -0.036551            -0.572700   0.383060   
Saudi Arabia              0.218319            -0.727003  -1.307234   
Hungary                   0.040904            -0.026706   0.152378   
Japan                    -0.478824             0.810089  -0.709266   

              urban_population  population_age_distribution_0_14  \
country                                                            
Vietnam              -1.445479                          0.717201   
Argentina             0.605509                          0.927114   
Saudi Arabia          0.310681                          0.962099   
Hungary              -0.144965                         -0.309038   
Japan                 0.597745                         -0.472303   

              population_age_distribution_60_plus  \
country                                             
Vietnam                                 -0.762264   
Argentina                               -0.437736   
Saudi Arabia                            -1.177358   
Hungary                                  0.362264   
Japan                                    0.920755   

              international_migrant_stock  restaurant_price_index_normal  
country                                                                   
Vietnam                         -0.678501                       3.210177  
Argentina                       -0.307692                       4.246653  
Saudi Arabia                     1.861933                       3.654267  
Hungary                         -0.323471                       3.745627  
Japan                           -0.560158                       4.186099

In [3]:
# iniciamos la ANOVA para la variable respuesta normalizada 'restaurant_price_index_normal'
lm = ols('''restaurant_price_index_normal ~ region + population_2017 + employment_agriculture + employment_industry + employment_services + education +
          quality_of_life_index + purchasing_power_index + safety_index + cost_of_living + consumer_price_index + human_capital_index +
          inflation + urban_population + population_age_distribution_0_14 + population_age_distribution_60_plus + international_migrant_stock''', 
          data=df).fit()

# resultados obtenidos en el ANOVA
sm.stats.anova_lm(lm)

df     sum_sq   mean_sq           F  \
region                               13.0  15.935057  1.225774   46.334942   
population_2017                       1.0   0.437499  0.437499   16.537694   
employment_agriculture                1.0   2.649718  2.649718  100.160859   
employment_industry                   1.0   0.235514  0.235514    8.902564   
employment_services                   1.0   0.253964  0.253964    9.599979   
education                             1.0   0.002535  0.002535    0.095834   
quality_of_life_index                 1.0   0.009480  0.009480    0.358344   
purchasing_power_index                1.0   0.059204  0.059204    2.237954   
safety_index                          1.0   0.019701  0.019701    0.744711   
cost_of_living                        1.0   2.798157  2.798157  105.771943   
consumer_price_index                  1.0   0.009834  0.009834    0.371714   
human_capital_index                   1.0   0.043091  0.043091    1.628860   
inflation                             1.0   0.155090  0.155090    5.862507   
urban_population                      1.0   0.168289  0.168289    6.361420   
population_age_distribution_0_14      1.0   0.000323  0.000323    0.012222   
population_age_distribution_60_plus   1.0   0.073129  0.073129    2.764306   
international_migrant_stock           1.0   0.008637  0.008637    0.326487   
Residual                             36.0   0.952367  0.026455         NaN   

                                           PR(>F)  
region                               1.682701e-18  
population_2017                      2.480687e-04  
employment_agriculture               6.080429e-12  
employment_industry                  5.089224e-03  
employment_services                  3.762664e-03  
education                            7.586718e-01  
quality_of_life_index                5.531754e-01  
purchasing_power_index               1.433751e-01  
safety_index                         3.938689e-01  
cost_of_living                       2.920061e-12  
consumer_price_index                 5.459009e-01  
human_capital_index                  2.100334e-01  
inflation                            2.063424e-02  
urban_population                     1.622612e-02  
population_age_distribution_0_14     9.125846e-01  
population_age_distribution_60_plus  1.050725e-01  
international_migrant_stock          5.712856e-01  
Residual                                      NaN

In [4]:
# sacamos el resumen de la recta que ajusta nuestro modelo
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                  OLS Regression Results                                 
=========================================================================================
Dep. Variable:     restaurant_price_index_normal   R-squared:                       0.960
Model:                                       OLS   Adj. R-squared:                  0.928
Method:                            Least Squares   F-statistic:                     29.80
Date:                           Sun, 21 May 2023   Prob (F-statistic):           1.28e-17
Time:                                   05:09:09   Log-Likelihood:                 46.219
No. Observations:                             66   AIC:                            -32.44
Df Residuals:                                 36   BIC:                             33.25
Df Model:                                     29                                         
Covariance Type:                       nonrobust                                         
=======================================================================================================
                                          coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------
Intercept                               3.2915      1.285      2.561      0.015       0.685       5.898
region[T.EasternAsia]                  -0.3965      0.251     -1.578      0.123      -0.906       0.113
region[T.EasternEurope]                 0.0755      0.223      0.338      0.737      -0.377       0.528
region[T.NorthernAfrica]                0.8769      0.400      2.193      0.035       0.066       1.688
region[T.NorthernAmerica]               0.1882      0.251      0.749      0.459      -0.322       0.698
region[T.NorthernEurope]                0.3087      0.228      1.354      0.184      -0.154       0.771
region[T.Oceania]                       0.0775      0.245      0.316      0.754      -0.420       0.575
region[T.South-easternAsia]            -0.3971      0.234     -1.695      0.099      -0.872       0.078
region[T.SouthAmerica]                 -0.0142      0.200     -0.071      0.944      -0.419       0.391
region[T.SouthernAfrica]                0.7228      0.315      2.297      0.028       0.085       1.361
region[T.SouthernAsia]                  0.1548      0.261      0.594      0.556      -0.374       0.683
region[T.SouthernEurope]                0.0726      0.235      0.309      0.759      -0.403       0.548
region[T.WesternAsia]                   0.1722      0.202      0.851      0.400      -0.238       0.582
region[T.WesternEurope]                 0.2381      0.240      0.992      0.328      -0.249       0.725
population_2017                         0.0097      0.007      1.306      0.200      -0.005       0.025
employment_agriculture                  3.1451      6.273      0.501      0.619      -9.578      15.868
employment_industry                     1.8602      3.732      0.498      0.621      -5.710       9.430
employment_services                     4.5391      9.039      0.502      0.619     -13.793      22.872
education                               0.0026      0.042      0.062      0.951      -0.083       0.089
quality_of_life_index                   0.0686      0.109      0.632      0.532      -0.152       0.289
purchasing_power_index                 -0.2469      0.096     -2.573      0.014      -0.442      -0.052
safety_index                           -0.0576      0.060     -0.954      0.347      -0.180       0.065
cost_of_living                          0.6364      0.079      8.070      0.000       0.476       0.796
consumer_price_index                    0.0007      0.027      0.025      0.980      -0.054       0.056
human_capital_index                     0.2659      0.106      2.499      0.017       0.050       0.482
inflation                              -0.0741    

CONCLUSIONES:

- El `R-Square` de nuestros datos es muy alto, llegando a 0.96, es decir, los datos sugieren que nuestro modelo de predición es muy fiable porque explica el 96% de la variación de la variable respuesta. Sin embargo, estas conclusiones son fictícias ya que nuestros datos no siguen una distribución normal. 
  
- En la columna P>|t| del summary vemos las variables que afectan la variable respuesta (pvalue <= 0.05) en nuetros datos son: `region', `purchasing_power_index`, `cost_of_living`, `human_capital_index`, `inflation` y `urban_population`. Estas son las que deberemos usar para el modelo de regresión lineal. 
  
- De las columnas que afectan nuestra VR sacamos los coef: `region` (con distintos valores), `purchasing_power_index` (coef: -0.2593), `cost_of_living` (coef 0.6338), `human_capital_index` (coef 0.2686), `inflation` (coef -0.0758) y `urban_population` (coef 0.2021). En el caso de la variable `purchasing_power_index` (coef: -0.2593), por cada unidad que aumenta `purchasing_power_index`, la variable respuesta disminuye 0.2593. 